libs/data-mapper/src/lib/mapDefinitions/MapDefinitionSerializer.ts

```diff
@@ -109,103 +113,151 @@ export const generateMapDefinitionBody = (mapDefinition: MapDefinitionEntry, con
-        const rootSourceNodes = rootTargetConnection.inputs[0];
-        const sourceNode = rootSourceNodes[0];
-        if (sourceNode && isConnectionUnit(sourceNode)) {
-          if (isFunctionData(sourceNode.node)) {
-            const valueToTrim = newPath
-              .map((pathItem) => (pathItem.key.startsWith(mapNodeParams.for) ? getSourceKeyOfLastLoop(pathItem.key) : ''))
-              .filter((path) => path !== '')
-              .join('/');
+      let formattedLmlSnippetForSource = createSourcePath(newPath, isFinalPath, targetPath, connections, input, pathToRoot);
+
+      // construct source side of LML for connection
+      if (isFinalPath) {
+        const connectionsToTarget = connections[addTargetReactFlowPrefix(targetPath.key)];
+        const inputIntoTargetNode = connectionsToTarget.inputs[0];
+        const inputNode = inputIntoTargetNode[0];
+        if (inputNode && isConnectionUnit(inputNode)) {
+          if (isFunctionData(inputNode.node)) {
+            const valueToTrim = getSrcPathRelativeToLoop(newPath);
 
             if (valueToTrim) {
               // Need local variables for functions
-              if (value === valueToTrim) {
-                value = '';
+              if (formattedLmlSnippetForSource === valueToTrim) {
+                formattedLmlSnippetForSource = '';
               } else {
-                value = value.replaceAll(`${valueToTrim}/`, '');
+                formattedLmlSnippetForSource = formattedLmlSnippetForSource.replaceAll(`${valueToTrim}/`, '');
 
                 // Handle dot access
-                if (!value.includes('[') && !value.includes(']')) {
-                  value = value.replaceAll(`${valueToTrim}`, '.');
+                if (!formattedLmlSnippetForSource.includes('[') && !formattedLmlSnippetForSource.includes(']')) {
+                  formattedLmlSnippetForSource = formattedLmlSnippetForSource.replaceAll(`${valueToTrim}`, '.');
                 }
               }
             }
           } else {
             // Need local variables for non-functions
-            const valueToTrim = findLast(sourceNode.node.pathToRoot, (pathItem) => pathItem.repeating && pathItem.key !== value)?.key;
-
-            if (value === sourceNode.node.key && sourceNode.node.nodeProperties.includes(SchemaNodeProperty.Repeating)) {
-              value = '.';
+            const valueToTrim = getPathForSrcSchemaNode(inputNode, formattedLmlSnippetForSource);
+            if (
+              formattedLmlSnippetForSource === inputNode.node.key &&
+              inputNode.node.nodeProperties.includes(SchemaNodeProperty.Repeating)
+            ) {
+              formattedLmlSnippetForSource = '.';
             } else if (valueToTrim) {
-              value = value.replace(`${valueToTrim}/`, '');
+              // account for source elements at different level of loop
+              let backoutValue = '';
+              if (valueToTrim !== lastLoop.loop && !valueToTrim.includes('/*')) {
```

Comment:
In which case would a part of data map value include '/*'?

Reply:
this is for json arrays! And there is a bug in the JSON array serialization, so this feature won't work for json for now